### ISSUE TITLE
Evaluation of External Methods

### ISSUE THREAD

**Issue by mahdip72** [rdilip/kanzi#2]
@rdilip Thanks for sharing this amazing work.

I have a question regarding the evaluation of external methods (e.g., ESM-3 VQVAE, FoldToken, etc.) reported in your paper. Did you rerun these methods yourself using official scripts or pipeline, or were the reported results taken directly from the original papers?

If you ran the evaluations yourself, could you please share any scripts or specific details about the evaluation setup?

**Comment by rdilip** [comment]
Thanks for your interest! We mostly reran these methods using the scripts present in each external repository to get reconstructed PDBs, since we test on a variety of datasets that aren't represented in all the original works. The setups for each are pretty different. I can share the metrics suite we made to compute things like designability and RMSD, though those are a little separate from the benchmarking (which was different for each method).

**Comment by mahdip72** [comment]
@rdilip Thanks for your response. 

I am the author of the [GCP-VQVAE](https://github.com/mahdip72/vq_encoder_decoder) paper. In our work, we evaluated three protein structure VQVAE methods. For this, we only considered backbone coordinates, omitting side chains and amino acid information. We focused on decoding the reconstructed representations produced by each method's VQ layer.

For example, in the case of ESM3 VQVAE, we could not find an official script for reconstruction, so we had to implement it ourselves.

I'm trying to understand the reason behind the discrepancy between your reported results and [ours](https://github.com/mahdip72/vq_encoder_decoder?tab=readme-ov-file#results) when evaluating two specific protein structure VQVAE methods: ESM3 VQVAE and Foldtoken4. 

Any help or guidance you can provide on these two methods would be a great favor and would help us strengthen our evaluation.

**Comment by rdilip** [comment]
@mahdip72 Ah, okay! Yes, I'll dig those up for you soon. From a cursory glance, though, it looks like you're reporting RMSDs of ~10A for ESM3's VQ-VAE, is that right? Let me know if I'm misreading something. That seems unusually high to me, and since that particular metric is just encode -> tokenize -> decode -> RMSD against original, so to my knowledge there isn't too much room for implementation specific issues.

(I am traveling atm so will need a few days to dig up the scripts we used for those -- I'll leave this issue open until I get to it)

**Comment by rdilip** [comment]
@mahdip72 I wrote up a quick snippet showing tokenization + detokenization. I can pretty consistently get close to the reported RMSD in the paper with this -- hope it helps!

Let me know if this answers your question.

```python

from abc import ABC
import torch
from pathlib import Path
from tqdm.auto import tqdm

from esm.models.esm3 import ESM3
from esm.pretrained import (
    ESM3_structure_encoder_v0, 
    ESM3_structure_decoder_v0, 
    ESM3_structure_decoder_v0,
)
class Tokenizer(ABC):
    def __init__(self):
        raise NotImplementedError

    def encode(self, coords_BLAD):
        """ returns idx_BL and s_BLD """
        raise NotImplementedError

    def decode(self, idx_BL):
        raise NotImplementedError

class ESMTokenizer(Tokenizer):
    def __init__(self, device):
        self.encoder = ESM3_structure_encoder_v0(device)
        self.decoder = ESM3_structure_decoder_v0(device)

        attached_model = ESM3.from_pretrained("esm3_sm_open_v1")
        self.bos = torch.tensor([[attached_model.tokenizers.structure.bos_token_id]]).to(device)
        self.eos = torch.tensor([[attached_model.tokenizers.structure.eos_token_id]]).to(device)
        self.device = device

    @torch.no_grad()
    def encode(self, coords_BLAD):
        assert coords_BLAD.size(0) == 1
        s, idx = self.encoder.encode(coords_BLAD.to(self.device))
        idx = torch.cat((self.bos, idx, self.eos), dim=-1)
        return idx, s

    @torch.no_grad()
    def decode(self, idx_BL):
        assert idx_BL.size(0) == 1
        
        if idx_BL[0, 0] != self.bos.item():
            idx_BL = torch.cat((self.bos, idx_BL, self.eos), dim=-1)

        recon = self.decoder.decode(idx_BL)['bb_pred'][:, 1:-1]
        return recon
```

**Comment by rdilip** [comment]
I'm closing this for now, but feel free to reach out if there are further questions!
